hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/snappy/SnappyDecompressor.java

```diff
@@ -276,13 +258,27 @@ public void end() {
     // do nothing
   }
 
-  private native static void initIDs();
+  private int decompressBytesDirect() throws IOException {
+    if (compressedDirectBufLen == 0) {
+      return 0;
+    } else {
+      // Set the position and limit of `compressedDirectBuf` for reading
+      compressedDirectBuf.position(0).limit(compressedDirectBufLen);
```

Comment:
I'm not sure we need to set position and limit here? If `compressedDirectBuf` already has been set with position and limit before calling `decompressBytesDirect`? Won't we read wrong data from this buffer?

Reply:
`decompressBytesDirect` is the only call that reads the data from `compressedDirectBuf`, so I think we should read it from the beginning. In `compressedDirectBuf`, we should fully decompress the  `compressedDirectBuf`.

Reply:
For `decompressDirect`, `compressedDirectBuf` is set up correctly before calling `decompressBytesDirect`. So we don't need to do it here again. For `decompress`, I'm not sure, but looks like `setInputFromSavedData` also takes care about `compressedDirectBuf` reseting. I guess we don't need to do it in `decompressBytesDirect`.